base/compiler/ssair/EscapeAnalysis/EscapeAnalysis.jl

```diff
@@ -0,0 +1,1604 @@
+    end
+
+    # if debug_itr_counter > 2
+    #     println("[EA] excessive iteration count found ", debug_itr_counter, " (", singleton_type(ir.argtypes[1]), ")")
+    # end
+
+    return estate
+end
+
+# propagate changes, and check convergence
+function propagate_changes!(estate::EscapeState, changes::Changes)
+    local anychanged = false
+    for change in changes
+        if isa(change, EscapeChange)
+            anychanged |= propagate_escape_change!(estate, change)
+        elseif isa(change, LivenessChange)
+            anychanged |= propagate_liveness_change!(estate, change)
+        else
+            change = change::AliasChange
+            anychanged |= propagate_alias_change!(estate, change)
+        end
+    end
+    return anychanged
+end
+
+@inline propagate_escape_change!(estate::EscapeState, change::EscapeChange) =
+    propagate_escape_change!(⊔, estate, change)
+
+# allows this to work as lattice join as well as lattice meet
+@inline function propagate_escape_change!(@nospecialize(op),
+    estate::EscapeState, change::EscapeChange)
+    (; xidx, xinfo) = change
+    anychanged = _propagate_escape_change!(op, estate, xidx, xinfo)
+    aliases = getaliases(xidx, estate)
+    if aliases !== nothing
+        for aidx in aliases
+            anychanged |= _propagate_escape_change!(op, estate, aidx, xinfo)
+        end
+    end
+    return anychanged
+end
+
+@inline function _propagate_escape_change!(@nospecialize(op),
+    estate::EscapeState, xidx::Int, info::EscapeLattice)
+    old = estate.escapes[xidx]
+    new = op(old, info)
+    if old ≠ new
+        estate.escapes[xidx] = new
+        return true
+    end
+    return false
+end
+
+# propagate Liveness changes separately in order to avoid constructing too many LivenessSet
+@inline function propagate_liveness_change!(estate::EscapeState, change::LivenessChange)
+    (; xidx, livepc) = change
+    info = estate.escapes[xidx]
+    Liveness = info.Liveness
+    Liveness === TOP_LIVENESS && return false
+    livepc in Liveness && return false
+    if Liveness === BOT_LIVENESS || Liveness === ARG_LIVENESS
+        # if this Liveness is a constant, we shouldn't modify it and propagate this change as a new EscapeLattice
+        Liveness = copy(Liveness)
+        push!(Liveness, livepc)
+        estate.escapes[xidx] = EscapeLattice(info; Liveness)
+        return true
+    else
+        # directly modify Liveness property in order to avoid excessive copies
+        push!(Liveness, livepc)
+        return true
+    end
+end
+
+@inline function propagate_alias_change!(estate::EscapeState, change::AliasChange)
+    (; xidx, yidx) = change
+    xroot = find_root!(estate.aliasset, xidx)
+    yroot = find_root!(estate.aliasset, yidx)
+    if xroot ≠ yroot
+        union!(estate.aliasset, xroot, yroot)
+        xinfo = estate.escapes[xidx]
+        yinfo = estate.escapes[yidx]
+        xyinfo = xinfo ⊔ yinfo
+        estate.escapes[xidx] = xyinfo
+        estate.escapes[yidx] = xyinfo
```

Comment:
does this need to happen for all elements of the aliasset, or why is it just the leaves of xidx and yidx?

Reply:
Now `add_alias_change!` will add new `EscapeChange` so that the `aliasset` equalization of newly added escape information will happen in `propagate_escape_change!`. It incurred some computational costs, but it's necessary for correctness.